sql,crosscluster: correctly handle NULLABLE columns in KV writer

stevendanna opened the pull request:
Previously, we were skipping encoding values based on whether the _new_ value was NULL, producing an incorrect _previous_ value.

Further, if _all_ columns were NULL we were not correctly writing a sentinel value.

Epic: none
Release note: None

stevendanna committed:
sql,crosscluster: correctly handle NULLABLE columns in KV writer

Previously, we were skipping encoding values based on whether the
_new_ value was NULL, producing an incorrect _previous_ value.

Further, if _all_ columns were NULL we were not correctly writing a
sentinel value.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131645)
<!-- Reviewable:end -->

msbutler commented on pkg/sql/row/deleter.go:
```diff
@@ -165,13 +165,15 @@ func (rd *Deleter) DeleteRow(
-			if prevValue.IsPresent() {
-				expValue = prevValue.TagAndDataBytes()
+			if !oth.PreviousWasDeleted {
+				prevValue, err := rd.encodeValueForPrimaryIndexFamily(family, values)
```
why is this extra logic around `oth.PreviousWasDeleted` only on the delete path, and not on the insert/update path?

Also, it's only possible for `!oth.PreviousWasDeleted && !prevValue.IsPresent()` if the prev value is null right?

stevendanna replied:
Good question, I'll write some words about this, maybe I overlooked something this morning.

Here is the issue.  A _parsed_ rangefeed event for a delete contains some of the columns. Namely, it contains any columns in the primary key.  This is also true for a _phantom delete_.  But, for a phantom delete, the correct expected value is nothing.  We need the columns from the key to construct the key, but when constructing the _expectation_ we want to make sure that we expect no key (as opposed to an existing key with a sentinel body).

msbutler replied:
yuck. given that the write will eventually succeed on a bogus expected value, i wonder it if is even worth having this extra logic. At the very least, perhaps we document this here?

msbutler commented on pkg/ccl/crosscluster/logical/lww_kv_processor_test.go:
```diff
@@ -0,0 +1,175 @@
+		return err
+	}
+
+	t.Run("one-NULL-to-not-NULL", func(t *testing.T) {
```
for all these test cases, if we randomly send a stale previous value, everything should just work, right? I.e. we could randomly choose to send kv1 instead of kv2 as the previous value.

stevendanna replied:
Yup, added some code that uses a random previous value about half of the time.

msbutler reviewed (commented):
Left only questions :D

stevendanna commented on pkg/sql/row/helper.go:
```diff
@@ -212,6 +214,42 @@ func (rh *RowHelper) encodeSecondaryIndexes(
 	return rh.indexEntries, nil
 }
 
+// encodePrimaryIndexValuesToBuf encodes the given values, writing
```
:D I'm still not sure I consider this "cleaned up", but it's a baby step.

stevendanna commented on pkg/sql/row/deleter.go:
```diff
@@ -212,35 +214,17 @@ func (rd *Deleter) encodeValueForPrimaryIndexFamily(
 	}
 	ret := roachpb.Value{}
-	if len(rd.rawValueBuf) > 0 {
+	if family.ID == 0 || len(rd.rawValueBuf) > 0 {
```
This is an area where we've made it a little hard to test because we don't support column families in a lot of our tooling. But here is my understanding. Consider the case where all of the not-null columns in the primary index are contained in the prety key and there are not composite-encoded columns in the primary key. In this case, the key contains all of the data we want to write. Here, there are two cases:

1. For family 0, our expected value should still contain a non-empty value. If it didn't it would look like a tombstone. So, we put in an empty tuple.

```
> select crdb_internal.pretty_key(key, -1), crdb_internal.pretty_value(value) from                                                            
                                -> crdb_internal.scan(crdb_internal.table_span('foo'::regclass::oid::int));                                                                    
  crdb_internal.pretty_key  | crdb_internal.pretty_value
----------------------------+-----------------------------
  /Tenant/2/Table/112/1/1/0 | /TUPLE/
```

2. For other column families, if there are no values to store into the family, then we don't expect a key to exist at all, so our expectation should be null.

msbutler replied:
every day i wake up grateful that i'm not a sql engineer.

michae2 replied:
😛

michae2 reviewed (commented):
> Reviewed 1 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131645)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @stevendanna)

___
*[`pkg/sql/row/writer.go` line 145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131645#-O8JSwpH8XIvR9gWVy8I:-O8JNFD-1Jj70JfoKONa:b-bn0gub) ([raw file](https://github.com/cockroachdb/cockroach/blob/99516bb0392f0dcc5e373a352d107aa61e36f0c2/pkg/sql/row/writer.go#L145)):*
> ```Go
> 			// Skip any values with a default ID not stored in the primary index,
> 			// which can happen if we are adding new columns.
> 			if skip := helper.SkipColumnNotInPrimaryIndexValue(family.DefaultColumnID, values[idx]); skip {
> ```

The fact that this skip depends on the value of the column means we also need to calculate skip separately for the old value and the new value.

Hmm... actually there's an existing bug here! This shouldn't always be `continue`. I've opened https://github.com/cockroachdb/cockroach/issues/131860 and have a fix going in https://github.com/cockroachdb/cockroach/pull/131869
___
*[`pkg/sql/row/writer.go` line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131645#-O8JmeVqA7bzA5C408Gy:-O8JmeVqA7bzA5C408Gz:baidcst) ([raw file](https://github.com/cockroachdb/cockroach/blob/99516bb0392f0dcc5e373a352d107aa61e36f0c2/pkg/sql/row/writer.go#L226)):*
> ```Go
> 				// set to NULL, so delete it.
> 				if oth.IsSet() {
> 					oth.DelWithCPut(ctx, batch, kvKey, expBytes, traceKV)
> ```

We only need to do this if `len(expBytes) > 0`. (If expBytes is zero, the previous KV shouldn't exist.)

Hmm, although now that I write that, I suppose we still need to issue the DelWithCPut to confirm that the previous KV doesn't exist and to check LWW. So scratch that.

Hmm! Which makes me realize, for this special LWW CPut mode, even if `family.ID != 0 && len(rawValueBuf) == 0 && !overwrite`, we probably still need to perform the DelWithCPut just in case there's a conflicting KV already there. So the condition above should be `if overwrite || oth.IsSet() {`.


<!-- Sent from Reviewable.io -->

michae2 commented:
(Sorry, I keep forgetting not to use Reviewable.)

stevendanna reviewed (commented):
:D No need to be sorry, feel free to use whatever you prefer.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131645#-:-O8LfAL36pBDBl9PgmGY:b-pvjjga)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @msbutler)

___
*[`pkg/sql/row/writer.go` line 145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131645#-O8JSwpH8XIvR9gWVy8I:-O8LfFgy-8zpUAtT8e0c:b-c49u6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/99516bb0392f0dcc5e373a352d107aa61e36f0c2/pkg/sql/row/writer.go#L145)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

The fact that this skip depends on the value of the column means we also need to calculate skip separately for the old value and the new value.

Hmm... actually there's an existing bug here! This shouldn't always be `continue`. I've opened https://github.com/cockroachdb/cockroach/issues/131860 and have a fix going in https://github.com/cockroachdb/cockroach/pull/131869
</blockquote></details>

Oh interesting. I didn't appreciate that composite encoding is value-dependent and not just type dependent.
___
*[`pkg/sql/row/writer.go` line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131645#-O8JmeVqA7bzA5C408Gy:-O8LfvwU71mBCzoP2a9T:b-2nuxrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/99516bb0392f0dcc5e373a352d107aa61e36f0c2/pkg/sql/row/writer.go#L226)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

We only need to do this if `len(expBytes) > 0`. (If expBytes is zero, the previous KV shouldn't exist.)

Hmm, although now that I write that, I suppose we still need to issue the DelWithCPut to confirm that the previous KV doesn't exist and to check LWW. So scratch that.

Hmm! Which makes me realize, for this special LWW CPut mode, even if `family.ID != 0 && len(rawValueBuf) == 0 && !overwrite`, we probably still need to perform the DelWithCPut just in case there's a conflicting KV already there. So the condition above should be `if overwrite || oth.IsSet() {`.
</blockquote></details>

Nice catch. It'll be a little hard to test this case end-to-end since LDR doesn't yet support multiple column families.


<!-- Sent from Reviewable.io -->

stevendanna committed:
sql: add assertion failure to prevent use of untested code path

We need to write tests for multi-column families as these will not be
hit by the end-to-end tests.

Epic: none
Release note: None

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131645)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @msbutler)

___
*[`pkg/sql/row/writer.go` line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131645#-O8JmeVqA7bzA5C408Gy:-O8MifDF7wHgQOMdnUrp:b-61lf8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/99516bb0392f0dcc5e373a352d107aa61e36f0c2/pkg/sql/row/writer.go#L226)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Nice catch. It'll be a little hard to test this case end-to-end since LDR doesn't yet support multiple column families.
</blockquote></details>

If fixed this up, but also: I added an assertion at the top of the file to make sure we can't use these multi-column family code paths until we do more thinking here.  


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 1 of 2 files at r3, 2 of 3 files at r4, 1 of 1 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131645#-:-O8OAYN48A2FfKvVydn_:bcmw7ss)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @msbutler)



<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r+
